### PR TITLE
chore(aws-api): move owner-based auth logic out of plugin

### PR DESCRIPTION
**Pull request by raphkim**
*Description of changes:*
Simple refactor to move the owner-based auth logic out of API plugin.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java [hunk 1]
```diff
@@ -276,68 +267,23 @@ public String getVersion() {
...
-            );
+            onSubscriptionFailure.accept(new ApiException(
+                    "No client information for API named " + apiName,
+                    "Check your amplify configuration to make sure there " +
+                            "is a correctly configured section for " + apiName
+            ));
             return null;
         }
 
         GraphQLRequest<R> request = graphQLRequest;
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/AWSApiPlugin.java:278]:
```suggestion
        GraphQLRequest<R> authDecoratedRequest = graphQLRequest;
```

File: aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+                    );
+                }
+                return oidcProvider.getLatestAuthToken();
+            case API_KEY:
+            case AWS_IAM:
+            case NONE:
+            default:
+                throw new ApiException(
+                        "Failed to obtain access token from the configured auth provider.",
+                        "Verify that the API is configured with either Cognito User Pools or OpenID Connect."
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java:218]: Should this say something about like `"@auth is only supported when suing Cognito User Pools or OpenID connect?"`

File: aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+                for (int i = 0; i < jsonGroups.length(); i++) {
+                    groups.add(jsonGroups.getString(i));
+                }
+            }
+        } catch (JSONException | CognitoParameterInvalidException error) {
+            throw new ApiException(
+                    "Failed to parse groups from auth rule.",
+                    error,
+                    "This should never happen - see attached exception for more details and report to us on GitHub."
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java:187]: There is the `REPORT_TO_AWS` (or whatever) message constant, can you use that?

File: aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        final String GROUPS_KEY = "cognito:groups";
+
+        try {
+            JSONObject accessToken = CognitoJWTParser.getPayload(getAuthToken(authType));
+
+            if (accessToken.has(GROUPS_KEY)) {
+                JSONArray jsonGroups = accessToken.getJSONArray(GROUPS_KEY);
+
+                for (int i = 0; i < jsonGroups.length(); i++) {
+                    groups.add(jsonGroups.getString(i));
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java:180]: Impossible to find single letter variables. I suggest:
```suggestion
                for (int index = 0; index < jsonGroups.length(); index++) {
                    groups.add(jsonGroups.getString(index));
```

File: aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            String value = getIdentityValue(idClaim, authType);
+
+            try {
+                return appSyncRequest.newBuilder()
+                        .variable(key, "String!", value)
+                        .build();
+            } catch (AmplifyException exception) {
+                throw new ApiException(
+                        "Failed to set owner field on AppSyncGraphQLRequest", exception,
+                        "See attached exception for details.");
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java:121]: Meh

File: aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+     * This will only work if the request is compliant with the AppSync specifications.
+     * @param request an instance of {@link GraphQLRequest}.
+     * @param authType the mode of authorization being used to authorize the request
+     * @param <R> The type of data contained in the GraphQLResponse expected from this request.
+     * @return the input request with additional variables that specify model's owner and/or
+     *          groups
+     * @throws ApiException If an error is encountered while processing the auth rules associated
+     *          with the request or if the authorization fails
+     */
+    public <R> GraphQLRequest<R> process(
```
Comment (jamesonwilliams) [inline aws-api/src/main/java/com/amplifyframework/api/aws/auth/AuthRuleProcessor.java:71]: maybe 
```suggestion
    public <R> GraphQLRequest<R> addAuthorizationFields(
```
or: 
```suggestion
    public <R> GraphQLRequest<R> decorate(
```
?

Should this be a `AuthRuleRequestDecorator`?

**Review by jamesonwilliams — COMMENTED** [review]
Definitely looks like a step in the right direction, thanks.

Now, we can write unit tests to verify that the decorator component is correctly generating outbound requests?

**Comment by richardmcclellan** [comment]
> Now, we can write unit tests to verify that the decorator component is correctly generating outbound requests?

I'd suggest moving the related tests in `AWSApiPluginTest` to a test class for this new decorator / processor.

**Review by jamesonwilliams — APPROVED** [review]
Ah, to watch software concerns be separated. There is no greater sight an engineer can see. 😍

**Pull request merged**